arch/risc-v/src/common/riscv_debug.c

```diff
@@ -0,0 +1,381 @@
+        {
+          slot = i;
+          break;
+        }
+    }
+
+  return slot;
+}
+
+static int riscv_debug_handler(int irq, FAR void *context, FAR void *arg)
+{
+  /* Get the trigger index */
+
+  int slot = READ_CSR(CSR_TSELECT);
+
+  DEBUGASSERT(slot >= 0);
+  DEBUGASSERT(slot < RISCV_DEBUG_NR_TRIGGER);
+
+  /* Call the trigger callback */
+
+  if (g_trigger_map[slot].callback)
+    {
+      g_trigger_map[slot].callback(g_trigger_map[slot].type,
+                                  g_trigger_map[slot].address,
+                                  g_trigger_map[slot].size,
+                                  g_trigger_map[slot].arg);
+    }
+
+  return OK;
+}
+
+/****************************************************************************
+ * Name: up_debugpoint_add
+ ****************************************************************************/
+
+void riscv_debug_init(void)
+{
+  union mcontrol mc;
+
+  /* Attach the debug exception handler */
+
+  irq_attach(RISCV_IRQ_BPOINT, riscv_debug_handler, NULL);
+
+  /* Clear the global trigger info & mark them as free slots */
+
+  memset(g_trigger_map, 0, sizeof(g_trigger_map));
+
+  /* Detect the support of NAPOT by trigger 0 */
+
+  WRITE_CSR(CSR_TSELECT, 0);
+
+  mc.reg = READ_CSR(CSR_TDATA1);
+
+  mc.match = 1;
+
+  /* Write it to tdata1 and read back
+   * to check if the NAPOT is supported
+   */
+
+  WRITE_CSR(CSR_TDATA1, mc.reg);
+  mc.reg = READ_CSR(CSR_TDATA1);
+
+  if (mc.match == 1)
+    {
+      g_support_napot = true;
+    }
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: up_debugpoint_add
+ ****************************************************************************/
+
+int up_debugpoint_add(int type, FAR void *addr, size_t size,
+                      debug_callback_t callback, FAR void *arg)
+{
+  int slot;
+  union mcontrol mc;
+  int ret = OK;
+  uintptr_t addr_napot;
+
+  /* Find a free slot */
+
+  slot = debug_find_slot(0, 0, 0);
+
+  if (slot >= 0)
+    {
+      /* Select the trigger */
+
+      WRITE_CSR(CSR_TSELECT, slot);
+
+      /* Fetch the current setting from tdata1 */
+
+      mc.reg = READ_CSR(CSR_TDATA1);
+
+      /* Configure trigger */
+
+      mc.m = 1;
+      mc.u = 1;
+      mc.hit = 0;
+      mc.dmode = 0;
+      mc.action = 0;
+
+      mc.execute = 0;
+      mc.load = 0;
+      mc.store = 0;
+
+      if (type == DEBUGPOINT_BREAKPOINT)
+        {
+          mc.execute = 1;
+        }
+      else if (type == DEBUGPOINT_WATCHPOINT_RO)
+        {
+          mc.load = 1;
+        }
+      else if (type == DEBUGPOINT_WATCHPOINT_WO)
+        {
+          mc.store = 1;
+        }
+      else if (type == DEBUGPOINT_WATCHPOINT_RW)
+        {
+          mc.load = 1;
+          mc.store = 1;
+        }
+      else
```

Comment:
why not handle DEBUGPOINT_STEPPOINT

Reply:
DEBUGPOINT_STEPPOINT is not implemented yet

Reply:
could we support it? gdb require this to work correctly.

Reply:
I do the test on QEMU and ESP32-C3, but both them does not support step point directly (they are not fully implemented the debug spec), to support it is another story by emulate it by breakpoint or debug mode.

Reply:
OK.